IntelliTect.TestTools.TestFramework/IntelliTect.TestTools.TestFramework/TestBuilder.cs

```diff
@@ -65,147 +83,89 @@ public TestBuilder AddDependencyService<T>()
+                        // Also these specific ones should not be overwriting TestBlockException
+                        var testBlockInstance = GetTestBlock(testBlockScope, fb.TestBlockType, logger);
+                        //if (TestBlockException != null) break;
 
-                        // Instead of doing this, might be worth extracting the above for loop into a private method and if that fails, then break out of the foreach we're in now
-                        if (testBlockException != null)
-                            break;
+                        SetTestBlockProperties(testBlockScope, ref testBlockInstance, logger);
+                        //if (TestBlockException != null) break;
 
-                        foreach (var arg in executeArgs)
-                        {
-                            logger.Debug($"Handing argument into Execute(): {GetObjectDataAsJsonString(arg)}");
-                        }
-                    }
+                        MethodInfo execute = GetExecuteMethod(testBlockScope, testBlockInstance);
+                        //if (TestBlockException != null) break;
 
-                    try
-                    {
-                        logger.Debug($"Running test block");
-                        var result = execute.Invoke(testBlockInstance, executeArgs);
-                        if (result != null)
-                        {
-                            logger.Debug($"Test block {tb.TestBlockType} returned... {GetObjectDataAsJsonString(result)}");
-                            testBlockResults.Add(result);
-                        }
+                        var executeArgs = GatherTestBlockArguments(testBlockScope, execute, fb, logger);
+                        //if (TestBlockException != null) break;
 
-                    }
-                    catch (TargetInvocationException ex)
-                    {
-                        logger.Error($"---Test block {tb.TestBlockType} failed.---");
-                        testBlockException = ex.InnerException;
-                        break;
-                    }
-                    catch (ArgumentException ex)
-                    {
-                        logger.Error($"---Test block {tb.TestBlockType} failed.---");
-                        testBlockException = ex;
-                        break;
-                    }
+                        RunTestBlocks(testBlockInstance, execute, executeArgs, logger);
+                        //if (TestBlockException != null) break;
 
-                    logger.Info($"---Test block {tb.TestBlockType} completed successfully.---");
+                    }
+                    TestBlockException = tempException;
                 }
             }
+            
 
             serviceProvider.Dispose();
 
-            if (testBlockException != null)
+            if (TestBlockException != null)
             {
-                throw testBlockException;
+                throw TestBlockException;
```

Comment:
This is probably not what you want here as it will smash the call stack on `TestBlockException`. What you want to do is wrap `TestBlockException` in another exception (probably pick a better exception type that this)
```
throw new Exception(TestBlockException);
```

Reply:
OH MY WORD, YES THANK YOU. I had forgot to ask someone about that... I remembered that there was some weirdness (to me) with rethrowing exceptions vs. throwing new ones like this, but I spaced dealing with it... Yes, this is perfect, it retains the ability for the tester to click on the stack trace in the test explorer to see where the test *really* failed, which is what I wanted! Thank you! I'll whip up... probably a custom exception for this and get it in today.